Fix nil pointer dereference when HTTP request fails

## Bug

In `main.go` line 98, the code accesses `res.StatusCode` before checking if `err` is nil:

```go
res, err := f.client.Do(req)
if res.StatusCode != 200 {   // BUG: res could be nil if err != nil
    return "", errors.New(res.Status)
}
```

When the HTTP request fails (timeout, network error), `res` is `nil` and `err` is non-nil. This causes a nil pointer dereference panic.

## Error observed in CI

**Failed run:** https://github.com/sensyn-robotics/crackmap/actions/runs/20323720225/job/58384656175

```
panic: template: .env:1:35: executing ".env" at <kv "https://...vault.azure.net/secrets/...">: 
error calling kv: runtime error: invalid memory address or nil pointer dereference
```

## Fix

```go
res, err := f.client.Do(req)
if err != nil {
    return "", err  // Check error FIRST
}
if res.StatusCode != 200 {
    return "", errors.New(res.Status)
}
```

## Additional improvements to consider

- Add retry logic for transient failures
- Replace `panic()` with graceful error handling
- Consider increasing timeout from 5 seconds

--- a/main.go
+++ b/main.go
@@ -104,6 +104,9 @@ func (f *fetcher) getToken() (string, error) {
 		req.Header.Add("Metadata", "true")
 	}
 	res, err := f.client.Do(req)
+	if err != nil {
+		return "", err
+	}
 	if res.StatusCode != 200 {
 		return "", errors.New(res.Status)
 	}

--- a/main_test.go
+++ b/main_test.go
@@ -3,6 +3,7 @@ package main
 import (
 	"bytes"
 	"errors"
+	"fmt"
 	"io/ioutil"
 	"net/http"
 	"os"
@@ -11,6 +12,12 @@ import (
 )
 
 type dummyClient struct{}
+
+type errorClient struct{}
+
+func (c *errorClient) Do(req *http.Request) (*http.Response, error) {
+	return nil, errors.New("connection timeout")
+}
 
 func (c *dummyClient) Do(req *http.Request) (*http.Response, error) {
 	var body string
@@ -132,4 +139,27 @@ PASSWORD=mysecretvalue1
 	if b.String() != expected {
 		t.Fatalf("got:%s want:%s", b.String(), expected)
 	}
+}
+
+func TestHttpRequestError(t *testing.T) {
+	var b bytes.Buffer
+	template := `PASSWORD={{ kv "https://example.vault.azure.net/secrets/pass" }}
+`
+	client := &errorClient{}
+	r := strings.NewReader(template)
+	defer func() {
+		if r := recover(); r != nil {
+			// Should panic with error message, not nil pointer dereference
+			panicMsg := fmt.Sprintf("%v", r)
+			if strings.Contains(panicMsg, "connection timeout") {
+				return // Expected behavior - error is properly propagated
+			}
+			if strings.Contains(panicMsg, "nil pointer") || strings.Contains(panicMsg, "invalid memory address") {
+				t.Fatalf("nil pointer dereference detected - fix not working: %v", r)
+			}
+			t.Fatalf("unexpected panic: %v", r)
+		}
+	}()
+	filter(fetcher{client, ""}, r, &b)
+	t.Fatalf("expected panic due to error, but none occurred")
 }

--- a/main_test.go
+++ b/main_test.go
@@ -148,16 +148,16 @@ func TestHttpRequestError(t *testing.T) {
 	client := &errorClient{}
 	r := strings.NewReader(template)
 	defer func() {
-		if r := recover(); r != nil {
+		if panicVal := recover(); panicVal != nil {
 			// Should panic with error message, not nil pointer dereference
-			panicMsg := fmt.Sprintf("%v", r)
+			panicMsg := fmt.Sprintf("%v", panicVal)
 			if strings.Contains(panicMsg, "connection timeout") {
 				return // Expected behavior - error is properly propagated
 			}
 			if strings.Contains(panicMsg, "nil pointer") || strings.Contains(panicMsg, "invalid memory address") {
-				t.Fatalf("nil pointer dereference detected - fix not working: %v", r)
+				t.Fatalf("nil pointer dereference detected - fix not working: %v", panicVal)
 			}
-			t.Fatalf("unexpected panic: %v", r)
+			t.Fatalf("unexpected panic: %v", panicVal)
 		}
 	}()
 	filter(fetcher{client, ""}, r, &b)

